react/ProfileForm.tsx

```diff
@@ -0,0 +1,475 @@
+    dispatch({
+      type: 'update',
+      field: evt.target.name as keyof ProfileState,
+      value: evt.target.value,
+      isValid: evt.target.value.length > 0,
+    })
+  }
+
+  const handleDocumentChange = ({
+    document,
+    documentType,
+    isValid,
+  }: {
+    document: string
+    documentType: string
+    isValid: boolean
+  }) => {
+    dispatch({
+      type: 'update',
+      field: 'document',
+      value: document,
+      isValid,
+    })
+
+    dispatch({
+      type: 'update',
+      field: 'documentType',
+      value: documentType,
+    })
+
+    if (!isValid) {
+      dispatch({
+        type: 'set_error',
+        field: 'document',
+        isValid: false,
+        error:
+          document.length > 0
+            ? messages.invalidDocumentMessage
+            : messages.fieldRequiredMessage,
+      })
+    }
+  }
+
+  const handleBlur: React.FocusEventHandler<HTMLInputElement> = evt => {
+    dispatch({
+      type: 'blur',
+      field: evt.target.name as keyof ProfileState,
+    })
+  }
+
+  const handlePhoneBlur: React.FocusEventHandler<HTMLInputElement> = () => {
+    // this timeout is needed because the phone field component focus
+    // the listbox button after the change handler, so we need to wait
+    // for the next tick when the focus has been potentially changed to it
+    setTimeout(() => {
+      if (
+        document.activeElement == null ||
+        // consider that the listbox is also our "input", so we won't trigger
+        // the error message if the user is selecting the country calling code
+        (document.activeElement !== phoneInputRef.current &&
+          !document.activeElement.matches('[data-reach-listbox-list=""]') &&
+          !(
+            phoneContainerRef.current!.compareDocumentPosition(
+              document.activeElement
+            ) & Node.DOCUMENT_POSITION_CONTAINED_BY
+          ))
+      ) {
+        dispatch({
+          type: 'blur',
+          field: 'phone',
+        })
+      }
+    }, 0)
+  }
+
+  const handleSubmit: React.FormEventHandler<HTMLFormElement> = async evt => {
+    evt.preventDefault()
+
+    setSubmitFailed(false)
+
+    if (Object.values(profileData).some(({ isValid }) => !isValid)) {
+      // send blur event to all fields so it's clear to the user which
+      // fields are in an invalid state.
+      Object.keys(profileData).forEach(field =>
+        dispatch({ type: 'blur', field: field as keyof ProfileState })
+      )
+      return
+    }
+
+    setLoading(true)
+
+    const profileDataValues: {
+      [field in keyof ProfileState]: ProfileState[field]['value']
+    } = Object.keys(profileData).reduce(
+      (profile, fieldName) => ({
+        ...profile,
+        [fieldName]: profileData[fieldName as keyof ProfileState].value,
+      }),
+      {} as any
+    )
+
+    const [
+      { success: profileUpdateSuccess },
+      { success: clientPreferencesUpdateSuccess },
+    ] = await Promise.all([
```

Comment:
Don't you need a `try...catch` block here?

Reply:
hmm, I think so. I'll add it